File: cabot_influxdb_client/script/test_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Temperature
import logging
from influxdb_client import InfluxDBClient, Point
from influxdb_client.client.write_api import SYNCHRONOUS

logging.basicConfig(level=logging.DEBUG)

USERNAME = "cabot"
PASSWORD = "cabot-influxdb"
TOKEN = "a54a87f7-73a0-4534-9741-ad7ff4e7d111"
ORGANIZATION_NAME = "cabot"
BUCKET_NAME = "grafana-test"
host = "localhost"
port = 8086

influxdb_url = f"http://{host}:{port}"
token = TOKEN
org = ORGANIZATION_NAME
bucket = BUCKET_NAME

client = InfluxDBClient(url=influxdb_url, token=token, org=org)
write_api = client.write_api(write_options=SYNCHRONOUS)

class TemperatureSubscriber(Node):

    def __init__(self):
        super().__init__('temperature_subscriber')
        self.subscription = self.create_subscription(
            Temperature,
            'temparature',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning
        self.get_logger().info("node ready")

    def listener_callback(self, msg):
        self.get_logger().info(f"Received temperature: {msg.temperature}")
        point = Point("temperature").tag("unit", "celsius").field("value", msg.temperature)
        try:
            write_api.write(bucket=bucket, org=org, record=point)
        except:
            pass

def main(args=None):
    rclpy.init(args=args)
    temperature_subscriber = TemperatureSubscriber()
    rclpy.spin(temperature_subscriber)
    temperature_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cabot_influxdb_client/src/test_client.cpp
#include "test_client.hpp"

const std::string USERNAME = "cabot";
const std::string PASSWORD = "cabot-influxdb";
const std::string TOKEN = "";
const std::string ORGANIZATION_NAME = "cabot";
const std::string BUCKET_NAME = "grafana-test";
const std::string INFLUXDB_URL = "http://localhost:8086";


TemperatureSubscriber::TemperatureSubscriber() : Node("temperature_subscriber"){
  subscription_ = create_subscription<sensor_msgs::msg::Temperature>(
      "/temperature", 10, std::bind(&TemperatureSubscriber::listener_callback, this, std::placeholders::_1));
  RCLCPP_INFO(this->get_logger(), "Node is ready");
}

void TemperatureSubscriber::listener_callback(const sensor_msgs::msg::Temperature::SharedPtr msg){
  RCLCPP_INFO(this->get_logger(), "Received temperature: %.2f", msg->temperature);

  influxdb::Point point("temperature");
  point.addTag("unit", "celsius");
  point.addField("value", msg->temperature);
  point.setTimestamp(std::chrono::system_clock::now());

  try{
    std::unique_ptr<influxdb::InfluxDB> idb(influxdb::InfluxDBFactory::Get(INFLUXDB_URL));
    idb->write(std::move(point));
  }catch(const std::exception& e){
    RCLCPP_ERROR(this->get_logger(), "Error writing to InfluxDB: %s", e.what());
  }
}

int main(int argc, char* argv[]){
  rclcpp::init(argc, argv);
  std::unique_ptr<TemperatureSubscriber> node_ptr(std::make_unique<TemperatureSubscriber>());
  std::shared_ptr<TemperatureSubscriber> node(std::move(node_ptr));
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

